Analysis/LeptonID/src/PIDutil.cc

```diff
@@ -0,0 +1,374 @@
+#include "PIDutil.h"
+#include <cmath>
+#include <set>
+#include <TF2.h>
+
+#include <marlin/VerbosityLevels.h>
+
+#include <EVENT/ReconstructedParticle.h>
+
+float getCorrectdEdx(float dEdx, Track *track)
+{
+    // dEdx correction stolen from the dEdx correction processor
+    double lambda = abs(atan(track->getTanLambda()) * 180.0 / M_PI);
+    std::vector<double> par = {0.970205, 0.0007506, 4.41781e-8, 5.8222e-8};
+
+    double f3 = 1 / (par[0] + par[1] * lambda + par[2] * pow(lambda, 2) + par[3] * pow(lambda, 3));
+
+    float dEdx_v2 = dEdx * f3;
+    return dEdx_v2;
+}
+
+LCObject *getRelated(LCObject *p, LCRelationNavigator const &nav, int &_cw, int &_tw)
+{
+    int wcut = 500;
+    auto weights = nav.getRelatedToWeights(p);
+    int max_ci = -1;
+    int max_ti = -1;
+    int max_cw = 0;
+    int max_tw = 0;
+    for (size_t i = 0; i < weights.size(); i++)
+    {
+        float w = weights[i];
+        // weights in permille
+        int cw = int(w) / 10000;
+        int tw = int(w) % 10000;
+        if (cw > max_cw)
+        {
+            max_cw = cw;
+            max_ci = i;
+        }
+        if (tw > max_tw)
+        {
+            max_tw = tw;
+            max_ti = i;
+        }
+    }
+    if (max_ci != max_ti)
+    {
+        // confusion, skip particle
+        return nullptr;
+    }
+    if (max_cw < wcut || max_tw < wcut)
+    {
+        // too messy, skip
+        return nullptr;
+    }
+
+    _cw = max_cw;
+    _tw = max_tw;
+    auto objects = nav.getRelatedToObjects(p);
+    return objects[max_ci];
+
+}
+
+WeightedPoints3D getWeightedPoints3D(const Cluster *clu, const LCCollection *PandoraClusters)
+{
+    auto hits = clu->getCalorimeterHits();
+    size_t n_hits = hits.size();
+    if (n_hits > 0) {
+        // we have a REC file and can use all details
+        std::vector<double> ehits, xhits, yhits, zhits;
+        ehits.reserve(n_hits);
+        xhits.reserve(n_hits);
+        yhits.reserve(n_hits);
+        zhits.reserve(n_hits);
+        for (const auto *hit: hits) {
+            ehits.push_back(hit->getEnergy());
+            auto pos = hit->getPosition();
+            xhits.push_back(pos[0]);
+            yhits.push_back(pos[1]);
+            zhits.push_back(pos[2]);
+        }
+        auto res = WeightedPoints3D(int(n_hits), ehits.data(), xhits.data(), yhits.data(), zhits.data());
+        return res;
```

Comment:
```suggestion
        return WeightedPoints3D(int(n_hits), ehits.data(), xhits.data(), yhits.data(), zhits.data());
```
No need to rob the compiler of an an easy opportunity to do RVO. ;)